### PR TITLE
Adngel smoke emitter update

### PR DESCRIPTION
**Pull request by Adngel**
This branch is a refactor of the code used by Smoke Emitters.

- Reference variables have been created to improve the readibility.
- Functions created to modularize some actions.
- namespace created to encapsulate the functions related to the smoke emitters.
- Replaced several calls of the deprecated "GetRandomControl" function.
- Replaced item pointer by item reference.
- fixed errors with the particle velocity.

Also added some new features
- restored the big bubbles effect using the OCB +2
- Added use for the variable ItemFlags[2], when steam emitter is underwater, ItemFlags[2] will be the radius it uses to spawn the bubles. (32 if by default, set it to 512 if you want bubles spawning from the tile)
- Implemented the ability to color the smoke_emitter using the editor object tint (smoke_emitter_black and smoke_emitter_white will keep ignoring the color tint).
- Removed the steam shot position displacement, for now on, the steam shot will spawn from the object root, in the direction of the cone red vertex. It can be moved to the tile sides keeping the Shift and Ctrl while moving in the editor.
- Added negative OCBs to  deactivate the damage of the steam shots (to simulate poison decorative effect like the VCI levels from TR5).
- Added the ability to free rotate them upwards and downwards. (Used only for steam shots effects).
- Added steam shot effects to the Smoke_emitter_black and Smoke_emitter_white

Note: You need update Tomb Editor (master branch) if you want to have the option of rotate smoke emitters pitch rotations. (If you don't, you still can use smoke emitters and rotate them via lua scripts).

### OCB and Item Flags:
- OCB 0: Default Smoke
- OCB negative : Steam shot won't harm Lara (new)
- OCB =! 0: Horizontal steam shot (edited from 8&)
- OCB + (16 * X): x is the number of frames it pauses between steam shots 
- OCB 111 (removed) originally moved the source 512 behind to spawn steam shot from the wall, now the same can be achieved just moving the cone object to the wall.

Example
OCB between 1 and 15 Steam shot is constant
OCB between 16 and 31 Steam shot will pause 1 frame
OCB between 480 and 495 (16 x 30) Steam shot will pause 30 frames (1 second)
OCB between 2400 and 2415 (16 x 150) Steam shot will pause 150 frames (5 seconds)

ItemFlags 
- item.ItemFlags[0]: Timer for the pause between steam shots
- item.ItemFlags[1]: Timer for the active steam shots
- item.ItemFlags[2]: Acceleration of the steam shot particles
- item.ItemFlags[3]: Bit flags of the Steam Emiter (bit 0 = No Damage) (new)
 
 In Underwater rooms
- OCB 0: Intermitent bubbles emission.
- OCB 1: Continuous bubbles emission.
- OCB +2: Use big bubbles. (new)

ItemFlags
- item.ItemFlags[0]: Count the number of bubles it has to spawn.
- item.ItemFlags[1]: Flag used to spawn a serie of bubbles with no delay.
- item.ItemFlags[2]: Radius of the bubbles spawn horizontal plane. (by default is 32).  (new)

### Things to do:

- [ ] Testing.
- [x] Revision.
- [x] Implementation of the free axis rotation in editor and engine. (Changes made in editor master branch).

### CODE REVIEW

File: TombEngine/Game/effects/effects.h [hunk 1]
```diff
@@ -160,6 +160,27 @@ struct Particle
 	int fxObj;
 	int roomNumber;
 	unsigned char nodeNumber; // ParticleNodeOffsetIDs enum.
+
+	// TODO: Clean above mess.
```
Comment (Stranger1992) [inline TombEngine/Game/effects/effects.h:164]: Is this comment still needed?

**Review by Stranger1992 — COMMENTED** [review]
I have had a look over the code and it seems good. I will now begin testing and report any findings.

**Comment by Stranger1992** [comment]
# QA and Testing

## SMOKE_EMITTER_WHITE (underwater)

- OCB 0: Intermittent bubbles emission. Result: Working ✅
- > item.ItemFlags[0]: Result: Working ✅
- > item.ItemFlags[1]: Result: Working ✅
- > item.ItemFlags[2]: Result: Working ✅
- OCB 1: Continuous bubbles emission. Result: Working ✅
- > item.ItemFlags[0]: Result: Working ✅
- > item.ItemFlags[1]: Result: Working ✅
- > item.ItemFlags[2]: Result: Working ✅
- OCB +2: Use big bubbles. (new) Result: Working ✅
- > item.ItemFlags[0]: Result: Working ✅
- > item.ItemFlags[1]: Result: Working ✅
- > item.ItemFlags[2]: Result: Working ✅

## SMOKE_EMITTER (Dry Room)

- OCB 0: Default Smoke Result: Working ✅
- > Also tested with SMOKE_EMITTER_WHITE. Result: Working. Tint ignored ✅
- OCB < 0: Steam shot won't harm Lara (new)
- > Also tested with SMOKE_EMITTER_WHITE. Result: Working. Tint ignored ✅
- OCB =! 0: Horizontal steam shot (edited from 8&)
- > Also tested with SMOKE_EMITTER_WHITE. Result: Working. Tint ignored ✅
- > Tested changes of rotation and emitter did shoot steam in the same direction as expected ✅
- OCB + (16 * X): x is the number of frames it pauses between steam shots. Result: Needs Revision 🤚.
- > If the acceleration is changed with itemFlag[2]: it seems to lose the direction it was originally emitting in. The video below shows that when I enter the volume to modify the itemFlag: it changes

https://github.com/MontyTRC89/TombEngine/assets/84292688/acb59f9e-9e2b-4895-a29a-f8c7a979d093

image

# Conclusion

Once the final point is addressed: I think you're good to go! What a fun PR!

**Comment by Stranger1992** [comment]
Further testing:

This one is hard to explain but the freeROT is not working with all angles:

In editor:

image

Ingame

image

Project:  [SmokeEmitterBug.zip](https://github.com/user-attachments/files/16037355/SmokeEmitterBug.zip) (can use any textures and the default tombengine wad

**Comment by Stranger1992** [comment]
> Further testing:
> 
> This one is hard to explain but the freeROT is not working with all angles:
> 
> In editor:
> 
> image
> 
> Ingame
> 
> image
> 
> Project: [SmokeEmitterBug.zip](https://github.com/user-attachments/files/16037355/SmokeEmitterBug.zip) (can use any textures and the default tombengine wad

Seems to be happening if the X rotation is 45, 90, degrees or 135 degrees.

image

 Negative values however work.
 
image

File: TombEngine/Game/effects/effects.h (hunk 1 above)
Comment (Sezzary) [reply TombEngine/Game/effects/effects.h:164]: Yes, the commented section is to replace the implementation above.

**Comment by Stranger1992** [comment]
@Adngel Hi! Have you got time to look at the above bugs that I encountered so further testing can commence. Thanks in advance.

**Comment by Stranger1992** [comment]
Retest:

* ## FreeROT correctly now working !

image

image

* ## Velocity changes correctly retains angle of emitter !

https://github.com/user-attachments/assets/abc8a26a-09c6-46a9-ab46-5c1eb9528f35

I think this is now good to go. Well done :D

**Review by Stranger1992 — APPROVED** [review]

**Comment by Stranger1992** [comment]
# Testing

Bug: In the below video: I have a node that activates all smoke emitters `OnEnter` . Once I exit they are all disabled `OnLeave` however going back into the node area too quickly results in the console being spammed with "hijacking sound xxxx" and the game freezes:

OnEnter: 
image

OnLeave:
image

Video

https://github.com/user-attachments/assets/455f9ade-674c-41dd-ae6e-30b57a05647e

**Comment by Stranger1992** [comment]
I have tested the above and (aside from the above bug that is unrelated) I think the PR is good to go! All functionality is working as expected. Well done @Adngel :D

**Pull request merged**